src/server/cmd/pachctl/cmd/compat.go

```diff
@@ -0,0 +1,638 @@
+		"stop job": "stop-job",
+		"restart datum": "restart-datum",
+		"list datum": "list-datum",
+		"inspect datum": "inspect-datum",
+		"get logs": "get-logs",
+		"create pipeline": "create-pipeline",
+		"update pipeline": "update-pipeline",
+		"inspect pipeline": "inspect-pipeline",
+		"extract pipeline": "extract-pipeline",
+		"edit pipeline": "edit-pipeline",
+		"list pipeline": "list-pipeline",
+		"delete pipeline": "delete-pipeline",
+		"start pipeline": "start-pipeline",
+		"stop pipeline": "stop-pipeline",
+		"inspect cluster": "inspect-cluster",
+		"debug dump": "debug-dump",
+		"debug profile": "debug-profile",
+		"debug binary": "debug-binary",
+		"debug pprof": "debug-pprof",
+		"delete all": "delete-all",
+	}
+
+	for newName, oldName := range simpleCompat {
+		compatCmd := &cobra.Command{}
+		*compatCmd = *findCommand(newName)
+
+		useSplit := strings.SplitN(compatCmd.Use, " ", 2)
+		if len(useSplit) == 2 {
+			compatCmd.Use = fmt.Sprintf("%s %s", oldName, useSplit[1])
+		} else {
+			compatCmd.Use = oldName
+		}
+
+		commands = append(commands, compatCmd)
+	}
+
+	// Helper types for organizing more complicated command compatibility
+	type RunFunc func(*cobra.Command, []string)
+	type CompatChanges struct {
+		Use string
+		Example string
+		Run func(*cobra.Command, RunFunc) RunFunc
+	}
+
+	// These helper functions will transform positional command-line args and
+	// pass-through to the new implementations of the command so we can maintain
+	// a single code path.
+	transformRepoBranch := func(cmd *cobra.Command, newRun RunFunc) func([]string) error {
+		return func(args []string) error {
+			var newArgs []string
+			var repo string
+
+			for i, arg := range args {
+				switch i % 2 {
+				case 0:
+					repo = arg
+				case 1:
+					newArgs = append(newArgs, fmt.Sprintf("%s@%s", repo, arg))
+					repo = ""
+				}
+			}
+
+			if repo != "" {
+				newArgs = append(newArgs, repo)
+			}
+
+			newRun(cmd, newArgs)
+			return nil
+		}
+	}
+
+	transformRepoBranchFile := func(cmd *cobra.Command, newRun RunFunc) func([]string) error {
+		return func(args []string) error {
+			var newArgs []string
+			var repoBranch string
+
+			for i, arg := range args {
+				switch i % 3 {
+				case 0:
+					repoBranch = arg
+				case 1:
+					repoBranch = fmt.Sprintf("%s@%s", repoBranch, arg)
+				case 2:
+					newArgs = append(newArgs, fmt.Sprintf("%s:%s", repoBranch, arg))
+					repoBranch = ""
+				}
+			}
+
+			if repoBranch != "" {
+				newArgs = append(newArgs, repoBranch)
+			}
+
+			newRun(cmd, newArgs)
+			return nil
+		}
+	}
+
+	// A few places pre-v1.9 accepted arguments of the form repo/branch -
+	// post-v1.9, we now expect repo@branch
+	transformRepoSlashBranch := func(cmd *cobra.Command, newRun RunFunc) func([]string) error {
+		return func(args []string) error {
+			var newArgs []string
+
+			for _, arg := range args {
+				newArgs = append(newArgs, strings.Replace(arg, "/", "@", 1))
+			}
+
+			newRun(cmd, newArgs)
+			return nil
+		}
+	}
```

Comment:
Ah, I got a bit used to this code style from javascript, although I wouldn't recommend using it everywhere.  There are a couple advantages here:

* `findCommand` captures `rootCmd`, so we don't have to pass it around everywhere
* It is useful to contain all compatibility code in this function so that it can be deleted in one chunk rather than searching to make sure that all the orphaned helper functions get removed.

I realize this isn't ideal for long-term supported code, but I'm considering this function to be a temporary quarantine during deprecation.  Let me know if you want me to hoist these out of `applyCommandCompat1_8`.

Reply:
Nah I'm cool with leaving this as is.